jumanji/environments/routing/multi_cvrp/env.py

```diff
@@ -0,0 +1,499 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Sequence, Tuple
+
+import chex
+import jax
+import jax.numpy as jnp
+import matplotlib
+from numpy.typing import NDArray
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.routing.multi_cvrp.generator import (
+    Generator,
+    UniformRandomGenerator,
+)
+from jumanji.environments.routing.multi_cvrp.reward import DenseReward, RewardFn
+from jumanji.environments.routing.multi_cvrp.types import (
+    Node,
+    Observation,
+    ObsVehicle,
+    PenalityCoeff,
+    State,
+    StateVehicle,
+    TimeWindow,
+)
+from jumanji.environments.routing.multi_cvrp.utils import (
+    DEPOT_IDX,
+    compute_distance,
+    compute_time_penalties,
+    create_action_mask,
+    max_single_vehicle_distance,
+)
+from jumanji.environments.routing.multi_cvrp.viewer import MultiCVRPViewer
+from jumanji.types import TimeStep, restart, termination, transition
+from jumanji.viewer import Viewer
+
+
+class MultiCVRP(Environment[State]):
+    """
+    Multi-Vehicle Routing Problems with Soft Time Windows (MVRPSTW) environment as described in [1].
+    We simplfy the naming to multi-agent capacitated vehicle routing problem (MultiCVRP).
+
+    - reward: jax array (float32)
+        this global reward is provided to each agent. The reward is equal to the negative sum of the
+        distances between consecutive nodes at the end of the episode over all agents. All time
+        penalties are also added to the reward.
+
+    - observation and state:
+        the observation and state variable types are defined in:
+        jumanji/environments/routing/multi_cvrp/types.py
+
+    [1] Zhang et al. (2020). "Multi-Vehicle Routing Problems with Soft Time Windows: A
+    Multi-Agent Reinforcement Learning Approach".
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        viewer: Optional[Viewer] = None,
+    ):
+        """
+        Instantiates a `MultiCVRP` environment.
+
+        Args:
+            generator: `Generator` whose `__call__` instantiates an environment instance.
+                Implemented options are [`UniformRandomGenerator`].
+                Defaults to `UniformRandomGenerator` with `num_customers=20` and `num_vehicles=2`.
+            reward_fn: `RewardFn` whose `__call__` method computes the reward of an environment
+                transition. The function must compute the reward based on the current state
+                and whether the environment is done.
+                Implemented options are [`DenseReward`]. Defaults to `DenseReward`.
```

Comment:
sparse as well, right?
```suggestion
                Implemented options are [`DenseReward`, `SparseReward`]. Defaults to `DenseReward`.
```